Python 3.11 deprecation warning: 'pipes' is deprecated and slated for removal in Python 3.13

Running the tests for my plugin [tox-py](https://github.com/adamchainz/tox-py) on Python 3.11b1 with tox 3.25.0, this warning is emitted:

```
tests/test_integration.py:8: in <module>
    import tox
.tox/py311/lib/python3.11/site-packages/tox/__init__.py:11: in <module>
    from . import exception
.tox/py311/lib/python3.11/site-packages/tox/exception.py:2: in <module>
    import pipes
/opt/hostedtoolcache/Python/3.11.0-beta.1/x64/lib/python3.11/pipes.py:68: in <module>
    warnings._deprecated(__name__, remove=(3, 13))
/opt/hostedtoolcache/Python/3.11.0-beta.1/x64/lib/python3.11/warnings.py:514: in _deprecated
    warn(msg, DeprecationWarning, stacklevel=3)
E   DeprecationWarning: 'pipes' is deprecated and slated for removal in Python 3.13
```

I checked and the `rewrite` branch doesn't use `pipes`, so not sure if this is worth actioning. I'm ignoring it in my test suite for now.

--- a/CONTRIBUTORS
+++ b/CONTRIBUTORS
@@ -1,3 +1,4 @@
+Adam Johnson
 Albin Vass
 Alex Grönholm
 Alexander Loechel

--- a/docs/changelog/2417.bugfix.rst
+++ b/docs/changelog/2417.bugfix.rst
@@ -1,0 +1,1 @@
+Avoid importing ``pipes`` on Python 3.3+ to avoid ``DeprecationWarning`` on Python 3.11 -- by :user:`adamchainz`

--- a/src/tox/action.py
+++ b/src/tox/action.py
@@ -1,7 +1,6 @@
 from __future__ import absolute_import, unicode_literals
 
 import os
-import pipes
 import signal
 import subprocess
 import sys
@@ -17,6 +16,11 @@ from tox.exception import InvocationError
 from tox.reporter import Verbosity
 from tox.util.lock import get_unique_file
 from tox.util.stdlib import is_main_thread
+
+if sys.version_info >= (3, 3):
+    from shlex import quote as shlex_quote
+else:
+    from pipes import quote as shlex_quote
 
 
 class Action(object):
@@ -89,7 +93,7 @@ class Action(object):
         """this drives an interaction with a subprocess"""
         cwd = py.path.local() if cwd is None else cwd
         cmd_args = [str(x) for x in self._rewrite_args(cwd, args)]
-        cmd_args_shell = " ".join(pipes.quote(i) for i in cmd_args)
+        cmd_args_shell = " ".join(shlex_quote(i) for i in cmd_args)
         stream_getter = self._get_standard_streams(
             capture_err,
             cmd_args_shell,

--- a/src/tox/config/__init__.py
+++ b/src/tox/config/__init__.py
@@ -44,9 +44,9 @@ from .parallel import ENV_VAR_KEY_PUBLIC as PARALLEL_ENV_VAR_KEY_PUBLIC
 from .parallel import add_parallel_config, add_parallel_flags
 from .reporter import add_verbosity_commands
 
-try:
+if sys.version_info >= (3, 3):
     from shlex import quote as shlex_quote
-except ImportError:
+else:
     from pipes import quote as shlex_quote
 
 

--- a/src/tox/exception.py
+++ b/src/tox/exception.py
@@ -1,6 +1,11 @@
 import os
-import pipes
 import signal
+import sys
+
+if sys.version_info >= (3, 3):
+    from shlex import quote as shlex_quote
+else:
+    from pipes import quote as shlex_quote
 
 
 def exit_code_str(exception_name, command, exit_code):
@@ -96,7 +101,7 @@ class MissingRequirement(Error):
         self.config = config
 
     def __str__(self):
-        return " ".join(pipes.quote(i) for i in self.config.requires)
+        return " ".join(shlex_quote(i) for i in self.config.requires)
 
 
 class BadRequirement(Error):

--- a/src/tox/venv.py
+++ b/src/tox/venv.py
@@ -1,7 +1,6 @@
 import codecs
 import json
 import os
-import pipes
 import re
 import sys
 from itertools import chain
@@ -18,6 +17,11 @@ from tox.util.lock import get_unique_file
 from tox.util.path import ensure_empty_dir
 
 from .config import DepConfig
+
+if sys.version_info >= (3, 3):
+    from shlex import quote as shlex_quote
+else:
+    from pipes import quote as shlex_quote
 
 #: maximum parsed shebang interpreter length (see: prepend_shebang_interpreter)
 MAXINTERP = 2048
@@ -194,7 +198,7 @@ class VirtualEnv(object):
 
         if path is None:
             raise tox.exception.InvocationError(
-                "could not find executable {}".format(pipes.quote(name)),
+                "could not find executable {}".format(shlex_quote(name)),
             )
 
         return str(path)  # will not be rewritten for reporting
@@ -534,7 +538,7 @@ class VirtualEnv(object):
                 # happens if the same environment is invoked twice
                 message = "commands[{}] | {}".format(
                     i,
-                    " ".join(pipes.quote(str(x)) for x in argv),
+                    " ".join(shlex_quote(str(x)) for x in argv),
                 )
                 action.setactivity(name, message)
                 # check to see if we need to ignore the return code

--- a/tests/unit/session/test_session.py
+++ b/tests/unit/session/test_session.py
@@ -1,5 +1,4 @@
 import os
-import pipes
 import sys
 import textwrap
 from threading import Thread
@@ -10,6 +9,11 @@ import tox
 from tox.exception import MissingDependency, MissingDirectory
 from tox.package import resolve_package
 from tox.reporter import Verbosity
+
+if sys.version_info >= (3, 3):
+    from shlex import quote as shlex_quote
+else:
+    from pipes import quote as shlex_quote
 
 
 def test_resolve_pkg_missing_directory(tmpdir, mocksession):
@@ -355,7 +359,7 @@ def test_command_prev_fail_command_skip_post_run(cmd, initproj, mock_venv):
             ___________________________________ summary ___________________________________{}
             ERROR:   py: commands failed
         """.format(
-            pipes.quote(sys.executable),
+            shlex_quote(sys.executable),
             "_" if sys.platform != "win32" else "",
         ),
     )

